rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+This design proposes **a common in-memory data representation**, **a way to
+translate that into TF feedables** (np.ndarray or EagerTensors) and **a set of
+APIs** each component can use to get both.
+
+![alt_text](20191017-tfx-standardized-inputs/overview.png)
+
+## Common in-memory data representation
+
+[Apache Arrow](https://arrow.apache.org/) will be used as the common in-memory
+data representation. Beam-based TFX components will accept
+<code>PCollection[pyarrow.[RecordBatch](https://arrow.apache.org/docs/python/data.html#record-batches)]</code>.
+
+Each logical data format will have its own encoding convention,
+[discussed](#logical-data-encoding-in-arrow) in the detailed design.
+
+We chose Apache Arrow because:
+
+*   It’s Expressive enough.
+    *   Lossless encoding of (conformant) tf.Example, tf.SequenceExample
+    *   Can encode structured data (proto)
+*   It’s a columnar format. It works well with common TFX workloads:
+    *   Column (feature)-wise analysis
+    *   Feed a batch of columns (features) to TensorFlow.
+*   It’s OSS friendly.
+    *   Community support for more storage format I/O (e.g. Apache Parquet)
+    *   Friendly to other OSS data formats, both in-memory and on disk (e.g.
+        Pandas)
+    *   Friendly to numpy / TF: many Arrow array types share the same memory
+        layout with numpy ndarrays and certain type of TF (composite) Tensors.
+*   TF neutral.
+    *   Leaves the possibility of supporting other ML libraries open.
+
+## Translation from Arrow to TF feedables
+
+The analogy to this is parsing tf.Examples into TF feedables -- extra
+information is needed in this translation because a
+[`Feature`](https://github.com/tensorflow/tensorflow/blob/abfba15cd9734cec7ecd3d0661b146fc251c842d/tensorflow/core/example/feature.proto#L76)
+can be converted to a Tensor, a SparseTensor or a
+[RaggedTensor](https://www.tensorflow.org/guide/ragged_tensor) depending on the
+[feature specs](https://github.com/tensorflow/tensorflow/blob/635e23a774936b5fe6fa3ef3cb6e54b55d93f324/tensorflow/python/ops/parsing_ops.py#L46-L49).
+Currently this extra information is implicitly contained in the pipeline schema
+(an instance of the
+[TFMD Schema](https://github.com/tensorflow/metadata/blob/master/tensorflow_metadata/proto/v0/schema.proto))
+proto.
+
+Similarly, an Arrow column can be translated to various TF feedables.
+[An extension to the pipeline schema](#tensorrepresentation) is proposed to for
+a user to express the intention for conversion.
+
+The conversion can be efficient (zero-copy) in certain cases. It is
+[discussed](#efficient-arrow-tensor-conversion) in the detailed design.
+
+## Standardized Inputs APIs
+
+We propose a set of APIs that TFX components will call, and need to be
+implemented for each of the supported combination of {physical, logical} format.
+
+```py
+class TFXIO(object):
+  """Abstract basic class of all Standardized TFX inputs API implementations."""
+  def __init__(
+      self, pipeline_env,
```

Comment:
There is no mention of pipeline_env elsewhere, can you briefly clarify?

Reply:
Removed as it's distracting